### PR TITLE
Support advanced indexing with boolean array for getitem

### PR DESCRIPTION
**Pull request by wkentaro**
Related to https://github.com/pfnet/chainer/issues/487

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1011,14 +1012,19 @@ cdef class ndarray:
         cdef Py_ssize_t i, j, offset, ndim, n_newaxes, n_ellipses, ellipsis
         cdef Py_ssize_t ellipsis_sizem, s_start, s_stop, s_step, dim, ind
         cdef vector.vector[Py_ssize_t] shape, strides
-        if not isinstance(slices, tuple):
+        if isinstance(slices, ndarray):
```
Comment (delta2323) [inline cupy/core/core.pyx]: Why this conditional branching is needed?

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -1011,14 +1012,19 @@ cdef class ndarray:
...
         else:
             slices = list(slices)
 
-        for s in slices:
-            if isinstance(s, ndarray):
-                raise ValueError('Advanced indexing is not supported')
+        # boolean array indexing
+        if isinstance(slices, ndarray) and slices.dtype == bool:
+            assert internal.vector_equal(self._shape, slices._shape)
+            return _boolean_array_indexing(self, slices)
```
Comment (delta2323) [inline cupy/core/core.pyx]: As advanced indexing has not been fully supported yet, we still need to raise `ValueError` for unsupported slice types.

File: cupy/core/core.pyx [hunk 3]
```diff
@@ -1834,6 +1840,28 @@ cdef _take_kernel_0axis = ElementwiseKernel(
...
+    'if (boolean_array) out[nth] = a',
+    'cupy_boolean_array_indexing_nth')
+
+
+cpdef ndarray _boolean_array_indexing(ndarray a, ndarray boolean_array):
+    cdef int i, n_true
+    nth = ndarray(boolean_array.shape, dtype=int)
+    n_true = 0
+    boolean_array = boolean_array.flatten()
+    for i in range(boolean_array.size):
```
Comment (delta2323) [inline cupy/core/core.pyx]: This implementation loops over the whole array, which takes O(array_size). I think [`scan`](https://github.com/pfnet/chainer/blob/master/cupy/core/core.pyx#L2563) is useful implement this feature. So could you try to improve the implementation?

File: cupy/core/core.pyx (hunk 1 above)
Comment (wkentaro) [inline cupy/core/core.pyx]: Because without this, `slices = list(slices)` is applied in `else` block.

**Comment by wkentaro** [comment]
I think I have done changing for the requests.

File: cupy/core/core.pyx (hunk 1 above)
Comment (delta2323) [inline cupy/core/core.pyx]: I see. Thanks

**Comment by delta2323** [comment]
Please add what we have implemented [here](https://github.com/pfnet/chainer/pull/1840/files?diff=unified#diff-bebe923819e806ae22e9923de58e954cR1031).

**Comment by wkentaro** [comment]
You mean the comment? I updated.

**Comment by delta2323** [comment]
> You mean the comment? I updated.

Yes

**Comment by yuyu2172** [comment]
I looked into your code and I found that it fails to work in a case when `slices` is one `numpy.ndarray` of type integer.

Here is an example.
```
>>> x = cupy.array([1,2,3,4])  
>>> i = np.array([1,2])        
>>> y = x[i]                   
Traceback (most recent call last):                                                                        
  File "script.py", line 7, in <module>                                                                   
    y = x[i]                                                                                              
  File "cupy/core/core.pyx", line 1069, in cupy.core.core.ndarray.__getitem__ (cupy/core/core.cpp:23335)  
  File "cupy/core/core.pyx", line 944, in cupy.core.core.ndarray.__iadd__ (cupy/core/core.cpp:20918)      
  File "cupy/core/elementwise.pxi", line 731, in cupy.core.core.ufunc.__call__ (cupy/core/core.cpp:43610) 
  File "cupy/core/elementwise.pxi", line 107, in cupy.core.core._preprocess_args (cupy/core/core.cpp:31973
)                                                                                                         
TypeError: Unsupported type <type 'list'>                                                                 
```

I suggest to leave `slices` as list and stop changing it to arrays in the beginning of `__getitem__`.

**Comment by yuyu2172** [comment]
Also, it would be better if you write tests in  `/tests/cupy_tests/core_tests/test_ndarray_adv_indexing.py`.

**Comment by wkentaro** [comment]
I think I've done.

File: tests/cupy_tests/core_tests/test_ndarray_indexing.py [hunk 4]
```diff
@@ -56,8 +52,6 @@
     {'shape': (10,), 'transpose': None, 'indexes': (slice(-5, -10, -1),)},
     {'shape': (10,), 'transpose': None, 'indexes': (slice(-5, -11, -1),)},
     {'shape': (10,), 'transpose': None, 'indexes': (slice(-5, -12, -1),)},
-    {'shape': (10,), 'transpose': None,
```
Comment (delta2323) [inline tests/cupy_tests/core_tests/test_ndarray_indexing.py]: Could you move restore this test case and move it to `test_ndarray_adv_indexing.py` as well?

**Comment by delta2323** [comment]
@yuyu2172 Thank you for reporting the failure.

Is it possible to add the failure case @yuyu2172 mentioned as an unit test? > @wkentaro

**Comment by yuyu2172** [comment]
I sent PR that adds the failing case to the test in #1934 few days ago.
Should I close it?

**Comment by wkentaro** [comment]
Added the test cases.

**Comment by delta2323** [comment]
@yuyu2172 Sorry, I missed #1934. Now we can support the PR in this PR, I would like to close #1934. Thank you for your cooperation !

**Comment by yuyu2172** [comment]
`IndexError` for unsupported dtype should be consistent with NumPy.

```python
a = np.array([1,2,3])
f = np.array([1., 2.])
a[f]
IndexError: arrays used as indices must be of integer (or boolean) type
```

**Comment by wkentaro** [comment]
> IndexError for unsupported dtype should be consistent with NumPy.

What do you mean?

**Comment by wkentaro** [comment]
@delta2323 Do you have any idea why CI testing fails as below? On my local pc, it passes.

```
$ PYTHONWARNINGS='module::DeprecationWarning' nosetests -a '!gpu,!slow' --with-doctest chainer_tests
............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................SSSSSSSSSSSSSSSSSS.....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................S....................................................................................................................................................................................................................................................................................................../home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/chainer/link.py:449: DeprecationWarning: Link.zerograds is deprecated. Use Link.cleargrads instead.
  DeprecationWarning)
/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/chainer/variable.py:243: DeprecationWarning: Variable.zerograd is deprecated. Use Variable.cleargard instead.
  DeprecationWarning)
....................................................................../home/travis/build.sh: line 57:  2381 Killed                  PYTHONWARNINGS='module::DeprecationWarning' nosetests -a '!gpu,!slow' --with-doctest chainer_tests


The command "PYTHONWARNINGS='module::DeprecationWarning' nosetests -a '!gpu,!slow' --with-doctest chainer_tests" exited with 137.
```

**Comment by wkentaro** [comment]
@yuyu2172 lgty?
https://github.com/pfnet/chainer/pull/1840/commits/04aed1e6100ada9adf6c0b7a6099f61a0d3d793c

**Comment by yuyu2172** [comment]
@wkentaro 
lgtm

**Comment by delta2323** [comment]
LGTM

**Comment by delta2323** [comment]
@wkentaro 

> Do you have any idea why CI testing fails as below? 

It seems the test was kill by Travis due to the excessive usage of memory. This PR does not seem the direct cause of this error.

**Comment by wkentaro** [comment]
@delta2323 Thanks.

**Pull request merged**